DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+3. If not, check if the argument has a copy constructor that converts it to the parameter type.
+
+In the first 2 situations, if the check is true for a specific overload and that overload
+is selected to be called, if any of the parameters have a matching copy constructor, it will get called,
+otherwise, standard copying is employed. In the third situation, the overload is matched based on the
+conversion resulted from the copy constructor, therefore it the copy constructor will surely get called:
+
+```d
+struct A
+{
+    int[] a;
+    @implicit this(ref shared A) immutable {}
+}
+
+void fun(const A a) {}                  // 1
+void fun(immutable A a) {}              // 2
+
+struct B
+{
+    int a[];
+    @implicit this(ref A) shared {}             // ctor1
+    @implicit this(ref A) const {}              // ctor2
+    @implicit this(ref shared A) shared {}      // ctor3
+    @implicit this(ref immutable A) const {}    // ctor4
+}
+
+void gun(shared A) {}    // 3
+void gun(const A) {}     // 4
+
+void main()
+{
+    A a;
+    fun(a);             // calls 1 - conversion to const match; standard copying
+
+    const A a1;
+    fun(a1);             // calls 1 - direct match; standard copying
+
+    shared A a2;
+    fun(a2);             // calls 2 - conversion with copy ctor match; call copy ctor on parameter
+
+    B b;
+    fun(b);              // error: both gun functions match, because copy
+                         // ctor matches both ctor1 and ctor2
+
+    const B b1;
+    fun(b1)               // calls 4 - direct match; standard copying
+
+    immutable B b2;
+    fun(b2);              // calls 4 - conversion to const match; call ctor4
+
+    shared B b3;
+    fun(b3);              // calls 4 - direct match; call ctor3
+}
+```
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    @implicit this(ref B another) immutable {}
+
+}
+
+void main()
+{
+    immutable A ia;
+    A a = ia;            // 1 - calls copy constructor
+
+    B bc;
+    B b = bc;            // 2 - b is evaluated to B.fun
+}
+```
+
+In situations where both the copy constructor and `alias this` are suitable
+to solve the assignment (1), the copy constructor takes precedence over `alias this`
+because it is considered more specialized (the sole purpose of the copy constructor is to
+create copies). However, if no copy constructor in the overload set matches the exact
+qualified types of the source and the destination, the `alias this` is preferred (2).
```

Comment:
This seems to conflict with (from section copy constructor call vs standard copying)
> Whenever an initialization by copying is encountered, a call to a copy constructor is tried
and if a matching copy constructor is found it is used, otherwise if no copy constructor is
suitable, standard copying is employed

1. When the overload set does not match, then a standard copy is performed.
1. When the overload set does not match and alias this is present, alias this is performed
1. A copy constructor is considered to be more specialized

Today, if I have an `alias this` and no post blit, the `alias this` does not take precedence:

```
struct A {
    int a;
    immutable(A) fun() {
        writeln("Yo!");
        return immutable A(7);
    }

    alias fun this;
}

void main() {
    A a;
    A b = a; // No Yo!
}
```

Reply:
In your example the alias this is never attempted because there are no undefined lookups (`A b = a` is something the compiler knows how to do). However, here is a slightly modified version of your code:

```d
struct A
{
    int[] a;
    immutable(A) fun()
    {
        writeln("Yo");
        return immutable A([7]);
    }

    alias fun this;
}

void main()
{
    A a;
    immutable A b = a;   // error: cannot implicitly convert expression a of type A to immutable(A)
}
```

If I'm not missing something this code should be forwarded to alias this, but it isn't, which IMO is a bug. The section about alias this was trying to state that in such situations copy construction takes precedence over alias this.